core/src/main/java/io/grpc/ResolvedServerInfoGroup.java

```diff
@@ -0,0 +1,109 @@
+/*
+ *
+ *  * Copyright 2016, Google Inc. All rights reserved.
+ *  *
+ *  * Redistribution and use in source and binary forms, with or without
+ *  * modification, are permitted provided that the following conditions are
+ *  * met:
+ *  *
+ *  *    * Redistributions of source code must retain the above copyright
+ *  * notice, this list of conditions and the following disclaimer.
+ *  *    * Redistributions in binary form must reproduce the above
+ *  * copyright notice, this list of conditions and the following disclaimer
+ *  * in the documentation and/or other materials provided with the
+ *  * distribution.
+ *  *
+ *  *    * Neither the name of Google Inc. nor the names of its
+ *  * contributors may be used to endorse or promote products derived from
+ *  * this software without specific prior written permission.
+ *  *
+ *  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *  * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *  * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *  * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *  * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *  * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *  * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+package io.grpc;
+
+import com.google.common.collect.ImmutableList;
+
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import javax.annotation.concurrent.Immutable;
+
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1770")
+@Immutable
+public final class ResolvedServerInfoGroup {
+  private final List<ResolvedServerInfo> resolvedServerInfoList;
+  private final Attributes attributes;
+
+  public ResolvedServerInfoGroup(List<ResolvedServerInfo> resolvedServerInfoList, Attributes attributes) {
+    this.resolvedServerInfoList = resolvedServerInfoList;
```

Comment:
Since this class is annotated as `@Immutable`, you need to make a defensive copy.

Also, to prevent handing out a mutable version to consumers that invoke `getResolvedServerInfoList`, you want to wrap the resulting copy in `Collections.unmodifiableList(...)`.